### PR TITLE
CLOUDP-67931: mongocli iam organizations|orgs apiKey(s)|apikey(s) whitelist|ipwhitelist create

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-67931

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

## Further comments

```zsh
./bin/mongocli iam orgs apikey whitelist create --apiKey 5f24084d8dbffa3ad3f25b73 --ip 93.144.26.189 -o ""
Created new whitelist entry(s).

```

```zsh
./bin/mongocli iam orgs apikey whitelist create --apiKey 5f24084d8dbffa3ad3f25b73 --ip 93.144.26.181 --cidr 93.144.26.154/32
{
  "results": [
    {
      "cidrBlock": "93.144.26.154/32",
      "created": "2020-08-04T10:16:51Z",
      "ipAddress": "93.144.26.154",
      "links": [
        {
          "rel": "self",
          "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/orgs/5e429e7706822c6eac4d5970/apiKeys/5f24084d8dbffa3ad3f25b73/whitelist/93.144.26.154"
        }
      ]
    },
    {
      "cidrBlock": "93.144.26.181/32",
      "created": "2020-08-04T10:16:51Z",
      "ipAddress": "93.144.26.181",
      "links": [
        {
          "rel": "self",
          "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/orgs/5e429e7706822c6eac4d5970/apiKeys/5f24084d8dbffa3ad3f25b73/whitelist/93.144.26.181"
        }
      ]
    }
  ],
  "links": [
    {
      "rel": "self",
      "href": "https://cloud-dev.mongodb.com/api/atlas/v1.0/orgs/5e429e7706822c6eac4d5970/apiKeys/5f24084d8dbffa3ad3f25b73/whitelist?pageNum=1\u0026itemsPerPage=100"
    }
  ],
  "totalCount": 2
}
```

### CODE REVIEW

File: internal/cli/organizations/apikeys/whitelist/create.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+			return opts.PreRunEOrg(opts.init)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return opts.Run()
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.apyKey, flag.APIKey, "", usage.APIKey)
+	cmd.Flags().StringVar(&opts.entryType, flag.Type, ipAddress, usage.OrganizationWhitelistType)
+	cmd.Flags().StringSliceVar(&opts.entries, flag.Entry, []string{}, usage.WhitelistEntry)
```
Comment (gssbzn) [inline internal/cli/organizations/apikeys/whitelist/create.go:104]: [nit] sorry fi this is an annoying change but looking at the command I think we can do better
how do you feel making it --ip 192.168.0.1 --cidr 192.168.0.1, this way we save the switch and let create both types at the same time

With this we should add a check that one of the lists shoudl not be empty (`if len(ips) == 0 && len(cidrs) == 0 return err`)
And makes your construction of the list a bit simpler (no switch) we create a list for ips, one for cidrs and then join them together

Comment (andreaangiolillo) [reply internal/cli/organizations/apikeys/whitelist/create.go:104]: yes, I like it. In addition we wouldn’t need the type flag anymore 👍

File: internal/cli/organizations/apikeys/whitelist/create.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func (opts *CreateOpts) init() error {
+	var err error
+	opts.store, err = store.New(config.Default())
+	return err
+}
+
+func (opts *CreateOpts) newWhitelistAPIKeysReq() ([]*atlas.WhitelistAPIKeysReq, error) {
+	var whitelistRep []*atlas.WhitelistAPIKeysReq
```
Comment (gssbzn) [inline internal/cli/organizations/apikeys/whitelist/create.go:48]: we know the final size, let's set it up so we don't need to resize the slice latter
`whitelistRep := make([]*atlas.WhitelistAPIKeysReq, 0, len(opts.ips) + len(opts.cidrs))`

File: internal/cli/organizations/apikeys/whitelist/create.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	if len(opts.ips) == 0 && len(opts.cidrs) == 0 {
+		return nil, fmt.Errorf("at least one between --ip and --cidr must be used")
+	}
+	for _, v := range opts.ips {
+		whitelist := atlas.WhitelistAPIKeysReq{
+			IPAddress: v,
+		}
+		whitelistRep = append(whitelistRep, &whitelist)
```
Comment (gssbzn) [inline internal/cli/organizations/apikeys/whitelist/create.go:56]: [non-blocking] you can also add the `&` when creating the instance, like `whitelist := &atlas.WhitelistAPIKeysReq{`
(same with the `cidrs` block

File: internal/usage/usage.go [hunk 4]
```diff
@@ -70,7 +70,7 @@ const (
 	MetricMode                      = "Atlas computes the current metric value as an average."
 	NotificationToken               = "Slack API token or Bot token or Flowdock personal API token." //nolint:gosec // This is just a message not a password
 	NotificationsChannelName        = "Slack channel name. Required for the SLACK notifications type."
-	APIKey                          = "Datadog API Key, Opsgenie API Key, VictorOps API key."
+	APIKey                          = "API Key."
```
Comment (gssbzn) [inline internal/usage/usage.go:73]: can we not loose the old description? we can rename the old to "AlertConfigAPIKey` if you want but I think is important to keep the old description as well as the new one

File: internal/usage/usage.go [hunk 5]
```diff
@@ -121,6 +121,8 @@ const (
 	PartitionFields                 = "Fields to use to partition data. You can specify up to two frequently queried fields to use for partitioning data."
 	ArchiveAfter                    = "Number of days that specifies the age limit for the data in the live Atlas cluster."
 	TargetProjectID                 = "Unique identifier of the project that contains the destination cluster for the restore job."
+	WhitelistIPEntry                = "IP address to be added to the whitelist for the API key."
```
Comment (gssbzn) [inline internal/usage/usage.go:124]: [nit] 
```suggestion
	WhitelistIPEntry                = "IP address to be whitelisted for a given API key."
```

File: internal/usage/usage.go [hunk 6]
```diff
@@ -121,6 +121,8 @@ const (
...
+	WhitelistCIDREntry              = "Whitelist entry in CIDR notation to be added for the API key."
```
Comment (gssbzn) [inline internal/usage/usage.go:125]: [nit]
```suggestion
	WhitelistCIDREntry              = "Whitelist entry in CIDR notation to be added for a given API key."
```

File: internal/cli/organizations/apikeys/whitelist/create.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+func (opts *CreateOpts) init() error {
+	var err error
+	opts.store, err = store.New(config.Default())
+	return err
+}
+
+func (opts *CreateOpts) newWhitelistAPIKeysReq() ([]*atlas.WhitelistAPIKeysReq, error) {
+	var whitelistRep []*atlas.WhitelistAPIKeysReq
+	if len(opts.ips) == 0 && len(opts.cidrs) == 0 {
+		return nil, fmt.Errorf("at least one between --ip and --cidr must be used")
```
Comment (gssbzn) [inline internal/cli/organizations/apikeys/whitelist/create.go:50]: [nit]
```suggestion
		return nil, fmt.Errorf("either --ip or --cidr must be set")
```

**Pull request merged**